datasets/flwr_datasets/partitioner/natural_id_partitioner.py

```diff
@@ -42,6 +47,61 @@ def _create_int_partition_id_to_natural_id(self) -> None:
             zip(range(len(unique_natural_ids)), unique_natural_ids)
         )
 
+    def _create_natural_id_to_int_partition_id(self) -> None:
+        """Create a mapping from unique client ids from dataset to int indices.
+
+        Natural ids come from the column specified in `partition_by`. This object is
+        inverse of the `self._partition_id_to_natural_id`. This method assumes that
+        `self._partition_id_to_natural_id` already exist.
+        """
+        self._natural_id_to_partition_id = {
+            value: key for key, value in self._partition_id_to_natural_id.items()
+        }
+
+    def _create_partition_id_to_indices(self) -> None:
+        """Create an assignment of indices to the partition indices."""
+        natural_ids = np.array(self.dataset[self._partition_by])
+        unique_natural_ids = self.dataset.unique(self._partition_by)
```

Comment:
Could we do this without having to load the whole dataset into memory?

Reply:
To make it work quicker we need to load the column specified by `self._partition_by` to the memory. It's required later `np.unique(natural_ids, return_inverse=True)` (line 96) which returns the alternative representation (ints instead of e.g. strings of this column and at the end it speeds things up).